### ISSUE TITLE
Is there an example of iterating the $appnode input parameter in the appcreds_sync hooks?

### ISSUE THREAD

**Issue by m2web** [ibm-apiconnect/devportal-addons#12]
I am attempting to get values out of the $appnode input parameter within an apic_app_delete($appnode, $data) hook. I see where it is logged:

```bash
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-06-30 14:52:37: WARNING: [pool www] child 35996 said into stderr: "WATCHDOG: [DEBUG] [appcreds_sync] appcreds_sync_apic_app_delete hook invoked for: O:31:&quot;Drupal\Core\Field\FieldItemList&quot;:8:{s:7:&quot;*list&quot;;a:1:{i:0;O:52:&quot;Drupal\Core\Field\Plugin\Field\FieldType\IntegerItem&quot;:8:{s:13:&quot;*....
```

However, when attempting to determine the datatype of the $appnode:

```php
\Drupal::logger('appcreds_sync')->debug("What is $appnode?: " . get_debug_type($appnode));
```

The following is in the log:

```bash
 [     nginx stderr]   587 80eab5:08cd01:a86b32 2022-06-30 17:46:31: [error] 1013#1013: *116258 FastCGI sent in stderr: "PHP message: Error: Object of class Drupal\node\Entity\Node could not be converted to string in /web/platforms/devportal-9.x-10.0.4-20211201-1711/sites/wsfg.int-dev.apiportal-intdev10.ws.wsfgrp.net/modules/appcreds_sync/appcreds_sync.module on line 87 #0 [internal function]: appcreds_sync_apic_app_delete()
 ```

Also, when attempting:

```php
 \Drupal::logger('appcreds_sync')->debug("What is $appnode?: " . gettype($appnode));
 ```

The following is in the log:

```bash
 [     nginx stderr]   587 80eab5:08cd01:a86b32 2022-06-30 17:51:22: [error] 1013#1013: *116316 FastCGI sent in stderr: "PHP message: Error: Object of class Drupal\node\Entity\Node could not be converted to string in /web/platforms/devportal-9.x-10.0.4-20211201-1711/sites/wsfg.int-dev.apiportal-intdev10.ws.wsfgrp.net/modules/appcreds_sync/appcreds_sync.module on line 87 #0 [internal function]: appcreds_sync_apic_app_delete()
```

I understand that this is more of a PHP issue, however, I am trying to use the $appnode param and am stumped. Thanks!

**Comment by chrisdudley** [comment]
The answer is in the error message:

Object of class Drupal\node\Entity\Node

that’s what it is, it’s a drupal Node entity, a quick Google should find you what that looks like (sorry on a phone so harder to do from here).

stackoverflow likely has lots of examples of using drupal node hooks, this isn’t anything specific to apic or the developer portal, it’s a standard drupal node entity in a standard drupal node hook.

I’m aware there is quite a learning curve with drupal, I remember facing it myself - it is climbable though :-)

please let us know if have further questions

**Comment by m2web** [comment]
Thanks Chris. As I understand, I will need to iterate the $appnode entity looking for a key:value pair. 

While I will continue digging, if you have any examples, please send them my way. Thanks again!

**Comment by chrisdudley** [comment]
It’s not a map, it’s an object so it likely has getters and setters.
What field are you after and I’ll see if I can find the right function?

**Comment by m2web** [comment]
Thanks Chris! This is very helpful when I know the weekend is gearing up 🙂!

The property I need is "something" like $appnode.app_credentials.client_id.

**Comment by chrisdudley** [comment]
Aha, it would be one of the more complicated fields ;-)

Here, try this, it will create a new array called `$credentials` that has all of the application's credentials in it - remember an application can have multiple sets of credentials.

```
            $appCredentials = $appnode->application_credentials_refs->referencedEntities();
            if ($appCredentials !== NULL && is_array($appCredentials)) {
              foreach ($appCredentials as $cred) {
                $credentials[] = [
                  'client_id' => $cred->client_id(),
                  'name' => $cred->name(),
                  'title' => $cred->title(),
                  'description' => $cred->summary(),
                ];
              }
            }
```           

If you're interested in why this code is so complicated, its due to how the data is stored in drupal. Credentials are a separate 'entity' that are cross-referenced to the application for performance reasons, it means we can really exploit the relational database.
The downside is when you want to access the data its a little more complicated - that `referencedEntities()` function above tells Drupal to actually expand the xref ID and replace it with the target object rather than having to go do another sql query ourselves. The credentials themselves are then also objects and have getters and setters for each field.
Its never simple :-)

**Comment by m2web** [comment]
Awesome! I am on holiday today, with the 4th of July on Monday here in the USA, making a long weekend 🙂 so I will get back to you Tuesday at the latest with the outcome.

Is there is documentation somewhere showing the data models along with class diagrams of what you explained above well?

**Comment by m2web** [comment]
Chris,
I have the following in the appcreds_sync_apic_app_delete hook:

```php
$appCredentials = $appnode->application_credentials_refs->referencedEntities();
if ($appCredentials !== NULL && is_array($appCredentials)) {
\Drupal::logger('appcreds_sync')->debug("appCredentials Array size: " . count($appCredentials));
      foreach ($appCredentials as $cred) {
              $credentials[] = [
	              'client_id' => $cred->client_id(),
	              'name' => $cred->name(),
	              'title' => $cred->title(),
	              'description' => $cred->summary(),
              ];
      }
}
```
Looking at the log, the appCredentials array is empty:

```bash
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-02 15:24:48: WARNING: [pool www] child 35996 said into stderr: "WATCHDOG: [DEBUG] [appcreds_sync] appCredentials Array size: 0 | user: anonymous ...."
```
Any other ideas? Thanks again for your help!

**Comment by chrisdudley** [comment]
It might not be a fully populated node, try this:
```
$fullNode = Node::load($appnode->id());
$appCredentials = $fullNode->application_credentials_refs->referencedEntities();
if ($appCredentials !== NULL && is_array($appCredentials)) {
\Drupal::logger('appcreds_sync')->debug("appCredentials Array size: " . count($appCredentials));
      foreach ($appCredentials as $cred) {
              $credentials[] = [
	              'client_id' => $cred->client_id(),
	              'name' => $cred->name(),
	              'title' => $cred->title(),
	              'description' => $cred->summary(),
              ];
      }
}
```

You'll need to add `use Drupal\node\Entity\Node;` to the top of your file too in order to import the Node class.

**Comment by m2web** [comment]
This line: 

```php
$appCredentials = $fullNode->application_credentials_refs->referencedEntities();
```

return the following in the log:

```bash
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 14:09:41: WARNING: [pool www] child 80194 said into stderr: "WATCHDOG: [ERROR] [php] Error: Call to a member function referencedEntities() on null in appcreds_sync_apic_app_delete() (line 121 of /web/platforms/devportal-9.x-10.0.4-20211201-1711/sites/wsfg.int-dev.apiportal-intdev10.ws.wsfgrp.net/modules/appcreds_sync/appcreds_sync.module) #0 [internal function]: appcreds_sync_apic_app_delete()"
```

Again, thank you for your help!

**Comment by chrisdudley** [comment]
that means that fullNode is null - so likely need to work backwards to see why.
Fairly hard for me to debug line by line remotely, i dont think thats very efficient!

You'll need to dump stuff to the log (e.g. send it through serialise first so that its a string and then you can see what you're dealing with.) dumping appnode is likely to be too large, but try appnode->id() to check that it has a valid node id, and then go from there.

**Comment by m2web** [comment]
The $appnode->id() getter call was returning a string of an integer such as 1779.

I used the following:

```php
if (is_string($appnode->nid)) {
    $debug = $appnode->nid;
}
    else {
    $debug = serialize($appnode->nid);
}
\Drupal::logger('appcreds_sync')->debug('%function hook with $appnode->nid: ' . $debug);
```

To log:

```bash
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "WATCHDOG: [DEBUG] [appcreds_sync] %function hook with $appnode-&gt;nid: O:31:&quot;Drupal\Core\Field\FieldItemList&quot;:8:{s:7:&quot;*list&quot;;a:1:{i:0;O:52:&quot;Drupal\Core\Field\Plugin\Field\FieldType\IntegerItem&quot;:8:{s:13:&quot;*definition&quot;;O:51:&quot;Drupal\Core\Field\TypedData\FieldItemDataDefinition&quot;:2:{s:18:&quot;*fieldDefinition&quot;;O:37:&quot;Drupal\Core\Field\BaseFieldDefinition&quot;:5:{s:7:&quot;*type&quot;;s:7:&quot;integer&quot;;s:9:&quot;*schema&quot;;N;s:10:&quot;*indexes&quot;;a:0:@s:17:&quot;*itemDefinition&quot;;r:4;s:13:&quot;*definition&quot;;a:6:{s:5:&quot;label&quot;;O:48:&quot;Drupal\Core\StringTranslation\TranslatableMarkup&quot;:3:{s:9:&quot;*string&quot;;s:2:&quot;ID&quot;;s:12:&quot;*arguments&quot;;a:0:@s:10:&quot;*options&quot;;a:0:@}s:9:&quot;read-only&quot;;b:1;s:8:&quot;provider&quot;;s:4:&quot;node&quot;;s:10:&quot;field_name&quot;;s:3:&quot;nid&quot;;s:11:&quot;entity_type&quot;;s:"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "4:&quot;node&quot;;s:6:&quot;bundle&quot;;N;}}s:13:&quot;*definition&quot;;a:2:{s:4:&quot;type&quot;;s:18:&quot;field_item:integer&quot;;s:8:&quot;settings&quot;;a:6:@s:8:&quot;unsigned&quot;;b:1;s:4:&quot;size&quot;;s:6:&quot;normal&quot;;s:3:&quot;min&quot;;s:0:&quot;&quot;;s:3:&quot;max&quot;;s:0:&quot;&quot;;s:6:&quot;prefix&quot;;s:0:&quot;&quot;;s:6:&quot;suffix&quot;;s:0:&quot;&quot;;}}s:9:&quot;*values&quot;;a:1:@s:5:&quot;value&quot;;s:4:&quot;1781&quot;;s:13:&quot;*properties&quot;;a:0:@s:7:&quot;*name&quot;;i:0;s:9:&quot;*parent&quot;;r:1;s:14:&quot;*_serviceIds&quot;;a:1:@s:16:&quot;typedDataManager&quot;;s:18:&quot;typed_data_manager&quot;;s:18:&quot;*_entityStorages&quot;;a:0:@s:20:&quot;*stringTranslation&quot;;N;}}s:11:&quot;*langcode&quot;;s:2:&quot;en&quot;;s:13:&quot;*definition&quot;;r:5;s:7:&quot;*name&quot;;s:3:&quot;nid&quot;;s:9:&quot;*parent&quot;;O:48:&quot;Drupal\Core\Entity\Plugin\DataType\EntityAdapter&quot"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: ";:8:{s:9:&quot;*entity&quot;;O:23:&quot;Drupal\node\Entity\Node&quot;:29:{s:10:&quot;in_preview&quot;;N;s:9:&quot;*values&quot;;a:44:{s:3:&quot;nid&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:4:&quot;1781&quot;;}}s:3:&quot;vid&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:4:&quot;1781&quot;;}}s:4:&quot;type&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:9:&quot;target_id&quot;;s:11:&quot;application&quot;;}}s:4:&quot;uuid&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:36:&quot;68ef1fe3-1597-4e2e-9c92-aa020a72d665&quot;;}}s:8:&quot;langcode&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:2:&quot;en&quot;;}}s:12:&quot;revision_uid&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:9:&quot;target_id&quot;;s:1:&quot;9&quot;;}}s:18:&quot;revision_timestamp&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;val"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "ue&quot;;s:10:&quot;1657041427&quot;;}}s:12:&quot;revision_log&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}s:16:&quot;revision_default&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;1&quot;;}}s:17:&quot;isDefaultRevision&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;1&quot;;s:6:&quot;status&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;1&quot;;}}s:3:&quot;uid&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:9:&quot;target_id&quot;;s:1:&quot;9&quot;;}}s:5:&quot;title&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:12:&quot;Test v10 App&quot;;}}s:7:&quot;created&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:10:&quot;1657041427&quot;;}}s:7:&quot;changed&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:10:&quot;1657041427&quot;;}}s:7:&quot;promote&quot;;a:1:{s:9:&quot;x-default&quot;"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: ";a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;0&quot;;}}s:6:&quot;sticky&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;0&quot;;}}s:16:&quot;default_langcode&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;1&quot;;}}s:29:&quot;revision_translation_affected&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;1&quot;;}}s:26:&quot;content_translation_source&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:3:&quot;und&quot;;}}s:28:&quot;content_translation_outdated&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:1:&quot;0&quot;;}}s:15:&quot;apic_catalog_id&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:36:&quot;3afac7d8-2274-427b-9dac-a2902d5ec087&quot;;}}s:15:&quot;apic_created_at&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:10:"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "&quot;1657041426&quot;;}}s:13:&quot;apic_hostname&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:51:&quot;https://apimgmt-intdev10.ws.wsfgrp.net/consumer-api&quot;;}}s:16:&quot;apic_provider_id&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:36:&quot;d7a952bf-4740-4eaf-8473-dbaf8293ff19&quot;;}}s:10:&quot;apic_state&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:7:&quot;enabled&quot;;}}s:12:&quot;apic_summary&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:2:@s:5:&quot;value&quot;;s:23:&quot;Testing v10 Module here&quot;;s:6:&quot;format&quot;;N;}}s:15:&quot;apic_updated_at&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:10:&quot;1657041426&quot;;}}s:8:&quot;apic_url&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:92:&quot;/consumer-api/apps/50f1d5cf-4098-40ad-b082-a664bd5f1dbd/68357944-107a-4a66-b2c4"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "-a10af692e0e3&quot;;}}s:23:&quot;application_client_type&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:12:&quot;confidential&quot;;}}s:28:&quot;application_consumer_org_url&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:55:&quot;/consumer-api/orgs/50f1d5cf-4098-40ad-b082-a664bd5f1dbd&quot;;}}s:28:&quot;application_credentials_refs&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:9:&quot;target_id&quot;;s:3:&quot;265&quot;;}}s:16:&quot;application_data&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:{s:5:&quot;value&quot;;s:949:&quot;a:15:{s:4:&quot;type&quot;;s:3:&quot;app&quot;;s:11:&quot;api_version&quot;;s:5:&quot;2.0.0&quot;;s:2:&quot;id&quot;;s:36:&quot;68357944-107a-4a66-b2c4-a10af692e0e3&quot;;s:4:&quot;name&quot;;s:12:&quot;test-v10-app&quot;;s:5:&quot;title&quot;;s:12:&quot;Test v10 App&quot;;s:7:&quot;summary&quot;;s:23:&quot;Testing v10 Module here&quot;;s:5:&quo"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "t;state&quot;;s:7:&quot;enabled&quot;;s:14:&quot;image_endpoint&quot;;s:101:&quot;https://wsfg.int-dev.apiportal-intdev10.ws.wsfgrp.net/wsfg/int-dev/modules/apic_app/images/app_15.png&quot;;s:15:&quot;lifecycle_state&quot;;s:10:&quot;production&quot;;s:10:&quot;created_at&quot;;s:24:&quot;2022-07-05T17:17:06.739Z&quot;;s:10:&quot;updated_at&quot;;s:24:&quot;2022-07-05T17:17:06.739Z&quot;;s:7:&quot;org_url&quot;;s:55:&quot;/consumer-api/orgs/50f1d5cf-4098-40ad-b082-a664bd5f1dbd&quot;;s:3:&quot;url&quot;;s:92:&quot;/consumer-api/apps/50f1d5cf-4098-40ad-b082-a664bd5f1dbd/68357944-107a-4a66-b2c4-a10af692e0e3&quot;;s:19:&quot;app_credential_urls&quot;;a:1:@i:0;s:141:&quot;/consumer-api/apps/50f1d5cf-4098-40ad-b082-a664bd5f1dbd/68357944-107a-4a66-b2c4-a10af692e0e3/credentials/2b4affa1-3a64-47bb-b09b-116aee56bd4f&quot;;s:9:&quot;client_id&quot;;s:32:&quot;e0554a207faf63c1abd8dc4fe6d0ea66&quot;;}&quot;;}}}s:19:&quot;application_enabled&quot;;a"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: ":1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:4:&quot;true&quot;;}}s:14:&quot;application_id&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:36:&quot;68357944-107a-4a66-b2c4-a10af692e0e3&quot;;}}s:27:&quot;application_lifecycle_state&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:10:&quot;PRODUCTION&quot;;}}s:16:&quot;application_name&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:5:&quot;value&quot;;s:12:&quot;test-v10-app&quot;;}}s:7:&quot;metatag&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}s:4:&quot;path&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:{i:0;a:1:@s:8:&quot;langcode&quot;;s:2:&quot;en&quot;;}}s:9:&quot;menu_link&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}s:17:&quot;application_image&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}s:29:&quot;application_lifecycle_pending&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}s:30:&quot;application_redirect_endpoi"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "nts&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}s:29:&quot;application_subscription_refs&quot;;a:1:{s:9:&quot;x-default&quot;;a:0:@}}s:9:&quot;*fields&quot;;a:0:@s:19:&quot;*fieldDefinitions&quot;;N;s:12:&quot;*languages&quot;;N;s:14:&quot;*langcodeKey&quot;;s:8:&quot;langcode&quot;;s:21:&quot;*defaultLangcodeKey&quot;;s:16:&quot;default_langcode&quot;;s:17:&quot;*activeLangcode&quot;;s:9:&quot;x-default&quot;;s:18:&quot;*defaultLangcode&quot;;s:2:&quot;en&quot;;s:15:&quot;*translations&quot;;a:1:{s:9:&quot;x-default&quot;;a:1:@s:6:&quot;status&quot;;i:1;}s:24:&quot;*translationInitialize&quot;;b:0;s:14:&quot;*newRevision&quot;;b:0;s:20:&quot;*isDefaultRevision&quot;;s:1:&quot;1&quot;;s:13:&quot;*entityKeys&quot;;a:4:@s:6:&quot;bundle&quot;;s:11:&quot;application&quot;;s:2:&quot;id&quot;;s:4:&quot;1781&quot;;s:8:&quot;revision&quot;;s:4:&quot;1781&quot;;s:4:&quot;uuid&quot;;s:36:&quot;68ef1fe3-1597-4e2e-9c92-aa020a72d665&quot;;s:25:&quo"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "t;*translatableEntityKeys&quot;;a:8:{s:5:&quot;label&quot;;a:1:@s:9:&quot;x-default&quot;;s:12:&quot;Test v10 App&quot;;s:8:&quot;langcode&quot;;a:1:@s:9:&quot;x-default&quot;;s:2:&quot;en&quot;;s:6:&quot;status&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;1&quot;;s:9:&quot;published&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;1&quot;;s:3:&quot;uid&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;9&quot;;s:5:&quot;owner&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;9&quot;;s:16:&quot;default_langcode&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;1&quot;;s:29:&quot;revision_translation_affected&quot;;a:1:@s:9:&quot;x-default&quot;;s:1:&quot;1&quot;;}s:12:&quot;*validated&quot;;b:0;s:21:&quot;*validationRequired&quot;;b:0;s:19:&quot;*loadedRevisionId&quot;;s:4:&quot;1781&quot;;s:33:&quot;*revisionTranslationAffectedKey&quot;;s:29:&quot;revision_translation_affected&quot;;s:37:&quot;*enforceRevisionTranslationAffected&quot;;a:0:@s:15"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: ":&quot;*entityTypeId&quot;;s:4:&quot;node&quot;;s:15:&quot;*enforceIsNew&quot;;N;s:12:&quot;*typedData&quot;;N;s:16:&quot;*cacheContexts&quot;;a:0:@s:12:&quot;*cacheTags&quot;;a:0:@s:14:&quot;*cacheMaxAge&quot;;i:-1;s:14:&quot;*_serviceIds&quot;;a:0:@s:18:&quot;*_entityStorages&quot;;a:0:@s:12:&quot;*isSyncing&quot;;b:0;}s:13:&quot;*definition&quot;;O:49:&quot;Drupal\Core\Entity\TypedData\EntityDataDefinition&quot;:1:{s:13:&quot;*definition&quot;;a:1:{s:11:&quot;constraints&quot;;a:2:{s:10:&quot;EntityType&quot;;s:4:&quot;node&quot;;s:6:&quot;Bundle&quot;;a:1:@i:0;s:11:&quot;application&quot;;}}}s:7:&quot;*name&quot;;N;s:9:&quot;*parent&quot;;N;s:14:&quot;*_serviceIds&quot;;a:0:@s:18:&quot;*_entityStorages&quot;;a:0:@s:20:&quot;*stringTranslation&quot;;N;s:19:&quot;*typedDataManager&quot;;N;}s:14:&quot;*_serviceIds&quot;;a:1:@s:16:&quot;typedDataManager&quot;;s:18:&quot;typed_data_manager&quot;;s:18:&quot;*_entityStorages&quot;;a:0:@s:"
[     nginx stdout]   587 80eab5:08cd01:a86b32 2022-07-05 17:17:12: "10.131.77.205, 10.131.171.35" apiportal-intdev10.ws.wsfgrp.net "POST /wsfg/int-dev/application/68357944-107a-4a66-b2c4-a10af692e0e3/delete HTTP/1.1" 303 518 1322 1324 "https://apiportal-intdev10.ws.wsfgrp.net/wsfg/int-dev/application/68357944-107a-4a66-b2c4-a10af692e0e3/delete" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:102.0) Gecko/20100101 Firefox/102.0" 0.802 "-"
[   php-fpm stderr]   588 80eab5:08cd01:af4a9d 2022-07-05 17:17:12: WARNING: [pool www] child 1330 said into stderr: "20:&quot;*stringTranslation&quot;;N;} | user: anonymous | uri: https://wsfg.int-dev.apiportal-intdev10.ws.wsfgrp.net/wsfg/int-dev/application/68357944-107a-4a66-b2c4-a10af692e0e3/delete | referer: https://apiportal-intdev10.ws.wsfgrp.net/wsfg/int-dev/application/68357944-107a-4a66-b2c4-a10af692e0e3/delete
"
```

I see the `client_id` key and the value of `e0554a207faf63c1abd8dc4fe6d0ea66`. 

Other than manually rebuilding the $appnode->nid log output into an actual hierarchy, what is the notation for the client_id value in the  $appnode->nid object?

**Comment by m2web** [comment]
Just wanting to follow up on this topic. Any ideas? Thanks in advance as I know you are all busy.

**Comment by m2web** [comment]
With Anu's help from IBM Support, here is what I did to get the client ID when a Portal Application is deleted.

First, instead of using the <module_name>_apic_app_delete($appnode, $data) hook, I used the <module_name>__apic_app_pre_delete(NodeInterface $node, $data) hook. Within this method, I do the following:

```php
if (isset($data['application_credentials_refs'])) { 
	foreach ($data['application_credentials_refs'] as $ref) {
	$cred = ApplicationCredentials::load($ref);
		if ($cred !== NULL) {
		     $credentials[] = [ 
		    'client_id' => $cred->client_id(),
		];}
	}
}
```

I then do the following to get the client Id out of the array: 

```php
$clientId = $credentials[0]['client_id'];
```

Of course, this is an issue when the application has multiple creds, but this worked.

Thanks again for the help!
